docs/proposals/20230706-yurtappoverrider.md

```diff
@@ -0,0 +1,334 @@
+### Goals
+1. Customize the workloads in different regions
+2. Implement GrayRelease through this
+3. Specify the registry of the image to adapt the edge network
+### Non-Goals/Future Work
+1. Optimize YurtAppSet(about status, patch, and replicas)
+2. Optimize YurtAppDaemon(about status)
+## Proposal
+### Inspiration
+Reference to the design of ClusterRole and ClusterRoleBinding.
+
+1. Considering the simplicity of customized rendering configuration, an incremental-like approach is used to implement injection, i.e., only the parts that need to be modified need to be declared, including image, replicas, and upgrade strategy. Therefore, it is reasonable to abstract these configurable fields into an Item. The design of Item refers to the design of VolumeSource in kubernetes.
+2. In order to inject item into the workloads, we should create a new CRD named YurtAppOverrider, which consist of items and patches. Items replace a set of configuration for matching nodepools.
+<img src = "../img/yurtappoverrider/Inspiration.png" width="600">
+3. Patch supports more advanced add, delete and replace operations, similar to kubectl's json patch. We can convert a patch struct into an API interface call.
+
+   ```shell
+   kubectl patch deployment xxx --type='json' --patch='[{"op": "replace", "path": "/spec/template/spec/containers/0/image", "value":"tomcat"}]'
+   ```
+### YurtAppOverrider API
+1. YurtAppOverrider needs to be bound to YurtAppSet/YurtAppDaemon.
+Considering that there are multiple Deployment/StatefulSet per nodepool, as shown below, it must be bound to YurtAppSet/YurtAppDaemon for injection. We use subject field to bind it to YurtAppSet/YurtAppDaemon.
+2. YurtAppOverrider is responsible for injection of entries. We only need to create a new YurtAppOverrider resource for all nodepools under a YurtAppSet/YurtAppDaemon resource.
+
+```go
+// ImageItem specifies the corresponding container and the claimed image
+type ImageItem struct {
+    // ContainerName represents name of the container
+    // in which the Image will be replaced
+    ContainerName string `json:"containerName"`
+    // ImageClaim represents the claimed image name
+    //which is injected into the container above
+    ImageClaim string `json:"imageClaim"`
+}
+
+// Item represents configuration to be injected.
+// Only one of its members may be specified.
+type Item struct {
+    // +optional
+    Image *ImageItem `json:"image,omitempty"`
+    // +optional
+    Replicas *int32 `json:"replicas,omitempty"`
+    // +optional
+    UpgradeStrategy *string `json:"upgradeStrategy,omitempty"`
```

Comment:
how about remove `UpgradeStrategy`? because only override `UpgradeStrategy` is not enough, other settings like   MaxUnavailable and MaxSurge should also be overrided.